File: packages/stats/gbstats/models/results.py
from typing import List, Literal, Optional, Tuple, Union

from pydantic.dataclasses import dataclass
import pandas as pd


# Internal results classes
@dataclass
class EffectMomentsResult:
    point_estimate: float
    standard_error: float
    pairwise_sample_size: int
    error_message: Optional[str]
    post_stratification_applied: bool


@dataclass
class Uplift:
    dist: str
    mean: float
    stddev: float


ResponseCI = Tuple[Optional[float], Optional[float]]


@dataclass
class TestResult:
    expected: float
    ci: ResponseCI
    uplift: Uplift
    errorMessage: Optional[str]


RiskType = Literal["absolute", "relative"]


@dataclass
class BayesianTestResult(TestResult):
    chanceToWin: float
    risk: List[float]
    riskType: RiskType


PValueErrorMessage = Literal[
    "NUMERICAL_PVALUE_NOT_CONVERGED",
    "ALPHA_GREATER_THAN_0.5_FOR_SEQUENTIAL_ONE_SIDED_TEST",
]


@dataclass
class PValueResult:
    p_value: Optional[float] = None
    p_value_error_message: Optional[PValueErrorMessage] = None


# Data classes for return to the back end
@dataclass
class RealizedSettings:
    postStratificationApplied: bool


@dataclass
class SingleVariationResult:
    users: Optional[float]
    cr: Optional[float]
    ci: Optional[ResponseCI]


@dataclass
class BanditResult:
    singleVariationResults: Optional[List[SingleVariationResult]]
    currentWeights: Optional[List[float]]
    updatedWeights: Optional[List[float]]
    bestArmProbabilities: Optional[List[float]]
    seed: int
    updateMessage: Optional[str]
    error: Optional[str]
    reweight: bool
    weightsWereUpdated: bool


@dataclass
class MetricStats:
    users: int
    count: int
    stddev: float
    mean: float


@dataclass
class BaselineResponse:
    cr: float
    value: float
    users: float
    denominator: Optional[float]
    stats: MetricStats


@dataclass
class PowerResponse:
    status: str
    errorMessage: Optional[str]
    firstPeriodPairwiseSampleSize: Optional[float]
    targetMDE: float
    sigmahat2Delta: Optional[float]
    priorProper: Optional[bool]
    priorLiftMean: Optional[float]
    priorLiftVariance: Optional[float]
    upperBoundAchieved: Optional[bool]
    scalingFactor: Optional[float]


@dataclass
class FrequentistTestResult(TestResult):
    pValue: Optional[float]
    pValueErrorMessage: Optional[PValueErrorMessage]


@dataclass
class BayesianVariationResponseIndividual(BayesianTestResult, BaselineResponse):
    realizedSettings: RealizedSettings
    power: Optional[PowerResponse]


@dataclass
class FrequentistVariationResponseIndividual(FrequentistTestResult, BaselineResponse):
    realizedSettings: RealizedSettings
    power: Optional[PowerResponse] = None


@dataclass
class SupplementalResults:
    cupedUnadjusted: Optional[
        BayesianVariationResponseIndividual | FrequentistVariationResponseIndividual
    ] = None
    uncapped: Optional[
        BayesianVariationResponseIndividual | FrequentistVariationResponseIndividual
    ] = None
    flatPrior: Optional[
        BayesianVariationResponseIndividual | FrequentistVariationResponseIndividual
    ] = None
    unstratified: Optional[
        BayesianVariationResponseIndividual | FrequentistVariationResponseIndividual
    ] = None
    noVarianceReduction: Optional[
        BayesianVariationResponseIndividual | FrequentistVariationResponseIndividual
    ] = None


@dataclass
class BayesianVariationResponse(BayesianVariationResponseIndividual):
    supplementalResults: Optional[SupplementalResults] = None


@dataclass
class FrequentistVariationResponse(FrequentistVariationResponseIndividual):
    supplementalResults: Optional[SupplementalResults] = None


@dataclass
class BaselineResponseWithSupplementalResults(BaselineResponse):
    supplementalResults: Optional[SupplementalResults] = None


VariationResponse = Union[
    BayesianVariationResponse,
    FrequentistVariationResponse,
    BaselineResponseWithSupplementalResults,
]

VariationResponseIndividual = Union[
    BayesianVariationResponseIndividual,
    FrequentistVariationResponseIndividual,
    BaselineResponse,
]


@dataclass
class DimensionResponseIndividual:
    dimension: str
    srm: float
    variations: List[VariationResponseIndividual]

    def to_df(self) -> pd.DataFrame:
        df = pd.DataFrame(self.variations)
        df["dimension"] = self.dimension
        df["srm"] = self.srm
        return df


@dataclass
class DimensionResponse:
    dimension: str
    srm: float
    variations: List[VariationResponse]

    def to_df(self) -> pd.DataFrame:
        df = pd.DataFrame(self.variations)
        df["variation"] = df.index
        df["dimension"] = self.dimension
        df["srm"] = self.srm
        return df


@dataclass
class ExperimentMetricAnalysisResult:
    unknownVariations: List[str]
    multipleExposures: float
    dimensions: List[DimensionResponse]


@dataclass
class ExperimentMetricAnalysis:
    metric: str
    analyses: List[ExperimentMetricAnalysisResult]


@dataclass
class MultipleExperimentMetricAnalysis:
    id: str
    results: List[ExperimentMetricAnalysis]
    banditResult: Optional[BanditResult]
    error: Optional[str]
    traceback: Optional[str]

File: packages/stats-ts/src/results.ts
export interface EffectMomentsResult {
  point_estimate: number;
  standard_error: number;
  pairwise_sample_size: number;
  error_message: string | null;
  post_stratification_applied: boolean;
}

export interface Uplift {
  dist: string;
  mean: number;
  stddev: number;
}

export type ResponseCI = [number | null, number | null];

export interface TestResult {
  expected: number;
  ci: ResponseCI;
  uplift: Uplift;
  errorMessage: string | null;
}

export type RiskType = "absolute" | "relative";

export interface BayesianTestResult extends TestResult {
  chanceToWin: number;
  risk: number[];
  riskType: RiskType;
}

export type PValueErrorMessage =
  | "NUMERICAL_PVALUE_NOT_CONVERGED"
  | "ALPHA_GREATER_THAN_0.5_FOR_SEQUENTIAL_ONE_SIDED_TEST";

export interface PValueResult {
  p_value?: number | null;
  p_value_error_message?: PValueErrorMessage | null;
}

export interface RealizedSettings {
  postStratificationApplied: boolean;
}

export interface SingleVariationResult {
  users: number | null;
  cr: number | null;
  variationVariances: number | null;
  ci: ResponseCI | null;
}

export interface BanditResult {
  singleVariationResults: SingleVariationResult[] | null;
  currentWeights: number[] | null;
  updatedWeights: number[] | null;
  bestArmProbabilities: number[] | null;
  seed: number;
  updateMessage: string | null;
  error: string | null;
  reweight: boolean;
  weightsWereUpdated: boolean;
}

export type Context = Record<string, unknown>;

export interface ContextualBanditResponse {
  context: Context;
  sampleSizePerVariation: number[] | null;
  variationMeans: number[] | null;
  variationVariances: number[] | null;
  updatedWeights: number[] | null;
  bestArmProbabilities: number[] | null;
  updateMessage: string | null;
  error: string | null;
}

/** Per-context observed data (sample moments), as opposed to posterior moments. */
export interface ContextualBanditContextSummary {
  context: Context;
  sampleSizePerVariation: number[] | null;
  sampleMeans: number[] | null;
  sampleVariances: number[] | null;
  updatedWeights: number[] | null;
  bestArmProbabilities: number[] | null;
  updateMessage: string | null;
  error: string | null;
}

/** JSON-serializable mapping from observed context attribute values to tree leaf id. */
export interface ContextualLeafMapEntry {
  context: Record<string, string>;
  leafId: number;
}

/** Container for per-context bandit results. */
export interface ContextualBanditNoTreeResult {
  attributes: string[];
  responses: ContextualBanditResponse[];
}

/** Tree fit output. `leafMap` uses tuple context keys until JSON serialization. */
export interface ContextualBanditResult extends ContextualBanditNoTreeResult {
  responsesContext: ContextualBanditContextSummary[];
  leafMap: unknown;
}

export interface MetricStats {
  users: number;
  count: number;
  stddev: number;
  mean: number;
}

export interface BaselineResponse {
  cr: number;
  value: number;
  users: number;
  denominator: number | null;
  stats: MetricStats;
}

export interface PowerResponse {
  status: string;
  errorMessage: string | null;
  firstPeriodPairwiseSampleSize: number | null;
  targetMDE: number;
  sigmahat2Delta: number | null;
  priorProper: boolean | null;
  priorLiftMean: number | null;
  priorLiftVariance: number | null;
  upperBoundAchieved: boolean | null;
  scalingFactor: number | null;
}

export interface FrequentistTestResult extends TestResult {
  pValue: number | null;
  pValueErrorMessage: PValueErrorMessage | null;
}

export interface BayesianVariationResponseIndividual
  extends BayesianTestResult,
    BaselineResponse {
  realizedSettings: RealizedSettings;
  power: PowerResponse | null;
}

export interface FrequentistVariationResponseIndividual
  extends FrequentistTestResult,
    BaselineResponse {
  realizedSettings: RealizedSettings;
  power?: PowerResponse | null;
}

export type VariationResponseIndividual =
  | BayesianVariationResponseIndividual
  | FrequentistVariationResponseIndividual
  | BaselineResponse;

export interface SupplementalResults {
  cupedUnadjusted?: VariationResponseIndividual | null;
  uncapped?: VariationResponseIndividual | null;
  flatPrior?: VariationResponseIndividual | null;
  unstratified?: VariationResponseIndividual | null;
  noVarianceReduction?: VariationResponseIndividual | null;
}

export interface BayesianVariationResponse
  extends BayesianVariationResponseIndividual {
  supplementalResults?: SupplementalResults | null;
}

export interface FrequentistVariationResponse
  extends FrequentistVariationResponseIndividual {
  supplementalResults?: SupplementalResults | null;
}

export interface BaselineResponseWithSupplementalResults
  extends BaselineResponse {
  supplementalResults?: SupplementalResults | null;
}

export type VariationResponse =
  | BayesianVariationResponse
  | FrequentistVariationResponse
  | BaselineResponseWithSupplementalResults;

export interface DimensionResponseIndividual {
  dimension: string;
  srm: number;
  variations: VariationResponseIndividual[];
}

export interface DimensionResponse {
  dimension: string;
  srm: number;
  variations: VariationResponse[];
}

export interface ExperimentMetricAnalysisResult {
  unknownVariations: string[];
  multipleExposures: number;
  dimensions: DimensionResponse[];
}

export interface ExperimentMetricAnalysis {
  metric: string;
  analyses: ExperimentMetricAnalysisResult[];
}

export interface MultipleExperimentMetricAnalysis {
  id: string;
  results: ExperimentMetricAnalysis[];
  banditResult: BanditResult | null;
  contextualBanditResult: ContextualBanditResult | null;
  error: string | null;
  traceback: string | null;
}
